### PR TITLE
[WIP] Make api lists ESI-based

### PR DESCRIPTION
**Pull request by cwisecarver**
This PR makes list views in the API use ESI. 

Instead of drawing out the full object in JSON the object is replaced by a varnish esi:include tag. This allows cache invalidation to work correctly because every part of the API now depends on a url for each individual object. 

#### Sample Data for http://localhost:8000/v2/nodes/?format=jsonapi&page[size]=3
```json
{
    "data":[
    <esi:include src="http://localhost:8000/v2/nodes/kxfsn/?format=jsonapi"/>,
    <esi:include src="http://localhost:8000/v2/nodes/bnscw/?format=jsonapi"/>,
    <esi:include src="http://localhost:8000/v2/nodes/r7myv/?format=jsonapi"/>
],
    "links":{
       "first":null,
        "last":"http://localhost:8000/v2/nodes/?format=jsonapi&page=115&page%5Bsize%5D=3",
        "prev":null,
        "next":"http://localhost:8000/v2/nodes/?format=jsonapi&page=2&page%5Bsize%5D=3",
        "meta":{"total":345,"per_page":3}}
}
```

#### Special Note
- For list esi:include tags the query parameters on the initial url are compared to a blacklist and appended to the esi url

  - For embed ESI:include tags the query parameters are **not** included because recursion

### CODE REVIEW

**Comment by brianjgeiger** [comment]
Are there any tests that can be done for this? It looks good otherwise, insofar as.

**Comment by cwisecarver** [comment]
@brianjgeiger The existing integration tests for varnish, that run on travis, will check nearly every combination of api url with esi enabled and esi disabled to make sure the responses match. They also check cache invalidation.

As far as unit tests, I didn't write any, technically, the generated string is invalid json. I could throw some regex against it to confirm what I think is happening is indeed happening but my opinion was that then I would have two problems. :wink: 

I spent the last 2 hours or so functionally testing it which did allow me to find several migrations that I hadn't run that were causing 500s.

**Comment by brianjgeiger** [comment]
Works for me.

File: api/base/serializers.py [hunk 1]
```diff
@@ -769,6 +775,25 @@ def invalid_embeds(self, fields, embeds):
...
+    def to_esi_representation(self, data):
+        href = None
+        query_params_blacklist = ['page[size]', 'format']
+        try:
+            href = data.get_absolute_url()
+        except:
+            representation = super(JSONAPISerializer, self).to_representation(data)
+            return representation
+        else:
+            if href and not href == '{}':
```
Comment (brianjgeiger) [inline api/base/serializers.py]: So, if get_absolute_url succeeds but sends back {} or None, there's nothing that gets returned. Is that the proper case? I think that would return a None into the list. Which might serialize fine, I'm not sure.

**Comment by brianjgeiger** [comment]
Okay, I could only think of one potential issue. Pass complete. :penguin:

File: api/base/serializers.py (hunk 1 above)
Comment (cwisecarver) [inline api/base/serializers.py]: @brianjgeiger You're right. I stole that from a previous implementation and was returning a SkipField in the else clause. I just added in a failsafe. I should also probably make that exception clause be tighter. I'll figure out what to watch for an change that too.

Comment (brianjgeiger) [inline api/base/serializers.py]: :+1:

**Comment by sloria** [comment]
Merged into `feature/instutution-autodeploy` per @cwisecarver 's request

File: api/base/serializers.py [hunk 2]
```diff
@@ -769,6 +775,27 @@ def invalid_embeds(self, fields, embeds):
...
+    def to_esi_representation(self, data):
+        href = None
+        query_params_blacklist = ['page[size]', 'format']
+        try:
+            href = data.get_absolute_url()
+        except AttributeError:
+            representation = super(JSONAPISerializer, self).to_representation(data)
+            return representation
+        else:
+            if href and not href == '{}':
```
Comment (sloria) [inline api/base/serializers.py]: Minor: `if href and href != '{}':` is a bit more readable.

**Comment by cwisecarver** [comment]
@sloria I'm going to move this back to WIP. I found a bug in it. I'm working on a fix now.

**Comment by sloria** [comment]
@cwisecarver Ah, saw your message after I had merged it locally. Go ahead and send the bugfix to develop when it's ready. I'll hold off on deploying this to staging.

**Pull request merged**